File: hearts-scala/src/main/scala/funar/hearts/Cards.scala
package funar.hearts

enum Suit {
  case Diamonds
  case Clubs
  case Spades
  case Hearts
}

object Suit {
  val all = Seq(Diamonds, Clubs, Spades, Hearts)
}

enum Rank {
  case Two
  case Three
  case Four
  case Five
  case Six
  case Seven
  case Eight
  case Nine
  case Ten
  case Jack
  case Queen
  case King
  case Ace
}

object Rank {
  given ordering: Ordering[Rank] = Ordering.by(_.ordinal)

  val all: Seq[Rank] = Seq(Two, Three, Four, Five, Six, Seven, Eight, Nine, Ten, 
                           Jack, Queen, King, Ace)
}

case class Card(suit: Suit, rank : Rank) {
  // Eine Karte kann nur eine andere Karte gleicher Farbe nach Wert schlagen
  def beats(other: Card): Option[Boolean] =
    if (this.suit == other.suit) {
      import Rank.ordering
      Some(Ordering[Rank].gt(this.rank, other.rank))
    } else
      None
}

object Card {

  def cartesianProduct[A, B](as: Seq[A], bs: Seq[B]): Seq[(A, B)] =
    as.map { a => bs.map { b => (a, b) }} flatten

  val deck : Seq[Card] = cartesianProduct(Suit.all, Rank.all).map { case (suit, rank) => Card(suit, rank) }
}

case class Player(name: String)

object Cards {

  opaque type Hand = Set[Card]
  object Hand {

    def make(cards: Iterable[Card]): Hand = cards.toSet
    
    val empty: Hand = Set.empty

    def isEmpty(hand: Hand): Boolean = hand.isEmpty

    def cards(hand: Hand): Seq[Card] = hand.toSeq

    def containsCard(hand: Hand, card: Card): Boolean =
      hand.contains(card)

    def removeCard(hand: Hand, card: Card): Hand =
      hand - card
  }

  type Trick = List[(Player, Card)]
  object Trick {
    val empty: Trick = List.empty

    def isEmpty(trick: Trick): Boolean = trick.isEmpty

    def cards(trick: Trick): Seq[Card] = trick.map(_._2)

    def add(trick: Trick, player: Player, card: Card) =
      (player, card) +: trick

    def leadingCard(trick: Trick): Card =
      trick.last._2

    def containsCard(trick: Trick, card: Card): Boolean =
      trick.contains(card)

    def pretty(trick: Trick) = {
      def prettyOne(player: Player, card: Card): String = 
        prettyCard(card) ++ " from " ++ prettyPlayer(player)
      def p(s: Trick): String =
        s match {
          case Nil => ""
          case Seq((player, card)) => prettyOne(player, card)
          case ((player, card)::rest) => prettyOne(player, card) ++ ", " ++ p(rest)
        }
      p(trick.reverse)
    }

    def prettyPlayer(player: Player): String = player.name

    def prettyCard(card: Card): String =
      card.rank.toString ++ " of " ++ card.suit.toString

  }

  opaque type Pile = Set[Card]
  object Pile {
    val empty: Pile = Set.empty

    def isEmpty(pile: Pile): Boolean = pile.isEmpty

    def addTrick(pile: Pile, trick: Trick): Pile =
      pile.union(Trick.cards(trick).toSet)

    def cards(pile: Pile): Seq[Card] = pile.toSeq
  }

}

File: hearts/src/Cards.hs
-- Das französische Blatt
module Cards(Suit(..), Rank(..), Card(..),
             deck, cardOrder,
             Player(..), 
             Hand, makeHand, emptyHand, isHandEmpty, containsCard, removeCard, handCards,
             Trick, emptyTrick,  trickEmpty, addToTrick, trickToList, listToTrick, leadingCardOfTrick,
             Pile, emptyPile, pileEmpty, pileAddTrick, pileCards,
             allSuits, allRanks)
where

import qualified Data.Set as Set
import Data.Set (Set)

import qualified Data.Map.Strict as Map
import Data.Map.Strict (Map, (!))

data Suit = Diamonds | Clubs | Spades | Hearts
  deriving (Show, Eq, Ord)

-- | Liste aller Farben
-- >>> length allSuits
-- 4
allSuits :: [Suit]
allSuits = [Spades, Hearts, Diamonds, Clubs]

data Rank = Two | Three | Four | Five | Six | Seven | Eight | Nine | Ten
         | Jack | Queen | King | Ace
  deriving (Show, Eq, Ord)

-- | Liste aller Werte
-- >>> length allRanks
-- 13
allRanks :: [Rank]
allRanks = [Two, Three, Four, Five, Six, Seven, Eight, Nine, Ten, 
            Jack, Queen, King, Ace]

-- | Spielkarte
data Card = Card { suit :: Suit, rank :: Rank }
  deriving (Show, Eq, Ord)

-- | Eine Karte kann nur eine andere Karte gleicher Farbe nach Wert schlagen
cardOrder :: Card -> Card -> Maybe Ordering
cardOrder c1 c2 =
  if suit c1 == suit c2
    then Just (compare (rank c1) (rank c2))
    else Nothing


-- | Kreuzprodukt aus zwei Listen
-- >>> cartesianProduct [1,2,3] [4,5,6]
-- [(1,4),(1,5),(1,6),(2,4),(2,5),(2,6),(3,4),(3,5),(3,6)]
cartesianProduct :: [a] -> [b] -> [(a, b)]
cartesianProduct list1 list2 =
  concat (map (\ el1 -> map (\ el2 -> (el1, el2)) list2) list1)

-- | Das ganze Kartenspiel
-- >>> length deck
-- 52
deck :: [Card]
deck = map (uncurry Card) (cartesianProduct allSuits allRanks)

-- | Karten, die jemand auf der Hand hält
newtype Hand = Hand { unHand :: Set Card }
  deriving (Eq, Show)

makeHand :: [Card] -> Hand
makeHand cards = Hand (Set.fromList cards)

handCards :: Hand -> [Card]
handCards (Hand set) = Set.toList set

-- | Ist die Hand leer?
isHandEmpty :: Hand -> Bool
isHandEmpty hand = Set.null (unHand hand)

-- | Enthält die Hand eine spezifische Karte?
containsCard :: Card -> Hand -> Bool
containsCard card hand = Set.member card (unHand hand)

-- | Karte aus der Hand entfernen
removeCard :: Card -> Hand -> Hand
removeCard card hand = Hand (Set.delete card (unHand hand))

-- | Leere Hand
emptyHand :: Hand
emptyHand = Hand Set.empty

-- Spieler

data Player = Player { playerName :: String }
  deriving (Show, Eq, Ord)

-- * Stich

-- | Zuletzt gespielte Karte zuerst
newtype Trick = Trick { trickToList :: [(Player, Card)] }
  deriving (Show, Eq)

listToTrick :: [(Player, Card)] -> Trick
listToTrick = Trick

-- | leeren Stich herstellen
emptyTrick :: Trick
emptyTrick = Trick []

-- | ist Stich leer
trickEmpty :: Trick -> Bool
trickEmpty (Trick list) = null list

-- | alle Karten desStich
cardsOfTrick :: Trick -> [Card]
cardsOfTrick (Trick list) = map snd list

-- | Karte auf den Stich legen
addToTrick :: Player -> Card -> Trick -> Trick
addToTrick player card (Trick list) = Trick ((player, card) : list)

-- | die Karte des Stich, die bedient werden muß
leadingCardOfTrick :: Trick -> Card
leadingCardOfTrick (Trick list) = snd (last list)

-- | Haufen aufgenommener Karten
newtype Pile = Pile { unPile :: Set Card }
  deriving Show

emptyPile :: Pile
emptyPile = Pile Set.empty

pileEmpty :: Pile -> Bool
pileEmpty pile = Set.null (unPile pile)

-- Stich auf den Haufen legen
pileAddTrick :: Pile -> Trick -> Pile
pileAddTrick pile trick =
  Pile (Set.union (unPile pile) (Set.fromList (cardsOfTrick trick)))

pileCards :: Pile -> [Card]
pileCards pile = Set.toList (unPile pile)

